core/src/main/scala/kafka/server/ReplicaManager.scala

```diff
@@ -1088,7 +1088,17 @@ class ReplicaManager(val config: KafkaConfig,
         quota = quota,
         clientMetadata = clientMetadata)
       if (isFromFollower) updateFollowerFetchState(replicaId, result)
-      else result
+      else {
+        // the fetch is from a consumer, sample the age of consumed data
+        if (random.nextDouble() <= config.liConsumerFetchSampleRatio) {
+          result.foreach { case (_, logReadResult) =>
+            logReadResult.info.records.batches().asScala.foreach { batch =>
+              ConsumerStats.consumedDataAgeHist.update(time.milliseconds - batch.maxTimestamp())
```

Comment:
Not all consumers have the same buffer size, do they? Couldn't this vary widely depending on which consumer it happens to be (in the sense of how many records/batches they're fetching)? I think it might be better to pick the first batch (or last one? average of the two?), which also might be a small perf win since it's O(1) rather than O(n).

Reply:
The fetch request has a maxSize field to control how much data should be returned.
Since we already have the config to specify the ratio of sampling, I prefer the simple logic here that samples all batches when the request is chosen.

Reply:
The problem is that you're not averaging or otherwise merging the samples from the different batches in a single request--they're all going into the histogram as separate data points, which means a one-batch update has only a tenth the influence of a 10-batch update. Nor is this a far-fetched example; we have many use cases where Samza jobs (Waterloo et al.) "grandfather" a topic from the beginning, and I'd bet they're using large buffers to maximize throughput--i.e., the samples are already artificially old due to the "read from the beginning" nature, and now you're multiplying the effect by the buffer size, roughly speaking.

Reply:
Understood. Let's compare two scenarios assuming two consumers are fetching from the same partition from a broker
- consumer1 using a maxSize of 1MB in Fetch requests;
- consumer2 using a maxSize of 10MB in Fetch requests.

To fetch 1000MB of data, consumer1 would have send 1000 requests and consumer2 100 requests. Assuming a sample ratio of 0.01, it means statistically consumer1 will have 10 requests chosen and consumer2 will only have 1 request chosen. If we go through all of the batches after a request is chosen, then consumer1 and consumer2 will have roughly the same number of batches contributing to the metric.

Reply:
I see, so yours is basically computing the per-byte-fetched average, while mine is computing the per-request average. I could see both as being useful... It was Serdar who was interested in this metric, wasn't it? Should we ask him?

Even if this is the only one chosen, I think it would be good to include "PerByte" or similar in the name so as to distinguish it from the other possibility.

Reply:
Per Serdar's feedback, he isn't strongly concerned with the details of how the statistic is computed, only that it's handled consistently across topics/etc.

Resolving thread.